src/pkg/backup/details/testdata/in_deets.go

```diff
@@ -0,0 +1,366 @@
+//	location set handling
+// ---------------------------------------------------------------------------
+
+var exists = struct{}{}
+
+type locSet struct {
+	// map [locationRef] map [itemRef] {}
+	// refs may be either the canonical ent refs, or something else,
+	// so long as they are consistent for the test in question
+	Locations map[string]map[string]struct{}
+	Deleted   map[string]map[string]struct{}
+}
+
+func newLocSet() *locSet {
+	return &locSet{
+		Locations: map[string]map[string]struct{}{},
+		Deleted:   map[string]map[string]struct{}{},
+	}
+}
+
+func (ls *locSet) AddItem(locationRef, itemRef string) {
+	if _, ok := ls.Locations[locationRef]; !ok {
+		ls.Locations[locationRef] = map[string]struct{}{}
+	}
+
+	ls.Locations[locationRef][itemRef] = exists
+	delete(ls.Deleted[locationRef], itemRef)
+}
+
+func (ls *locSet) RemoveItem(locationRef, itemRef string) {
+	delete(ls.Locations[locationRef], itemRef)
+
+	if _, ok := ls.Deleted[locationRef]; !ok {
+		ls.Deleted[locationRef] = map[string]struct{}{}
+	}
+
+	ls.Deleted[locationRef][itemRef] = exists
+}
+
+func (ls *locSet) AddLocation(locationRef string) {
+	if _, ok := ls.Locations[locationRef]; !ok {
+		ls.Locations[locationRef] = map[string]struct{}{}
+	}
+	// don't purge previously deleted items, or child locations.
+	// Assumption is that their itemRef is unique, and still deleted.
+	delete(ls.Deleted, locationRef)
+}
+
+func (ls *locSet) RemoveLocation(locationRef string) {
+	ss := ls.Subset(locationRef)
+
+	for lr := range ss.Locations {
+		items := ls.Locations[lr]
+
+		delete(ls.Locations, lr)
+
+		if _, ok := ls.Deleted[lr]; !ok {
+			ls.Deleted[lr] = map[string]struct{}{}
+		}
+
+		for ir := range items {
+			ls.Deleted[lr][ir] = exists
+		}
+	}
+}
+
+// MoveLocation takes the LAST elemet in the fromLocation (and all)
+// children matching the prefix, and relocates it as a child of toLocation.
+// ex: MoveLocation("/a/b/c", "/d") will move all entries with the prefix
+// "/a/b/c" into "/d/c".  This also deletes all "/a/b/c" entries and children.
+// assumes item IDs don't change across the migration.  If item IDs do change,
+// that difference will need to be handled manually by the caller.
+// returns the base folder's new location (ex: /d/c)
+func (ls *locSet) MoveLocation(fromLocation, toLocation string) string {
+	ss := ls.Subset(fromLocation)
+
+	fromBuilder := path.Builder{}.Append(path.Split(fromLocation)...)
+	toBuilder := path.Builder{}.Append(path.Split(toLocation)...).Append(fromBuilder.LastElem())
+
+	for lr, items := range ss.Locations {
+		lrBuilder := path.Builder{}.Append(path.Split(lr)...)
+		lrBuilder.UpdateParent(fromBuilder, toBuilder)
+
+		newLoc := lrBuilder.String()
+
+		for ir := range items {
+			ls.RemoveItem(lr, ir)
+			ls.AddItem(newLoc, ir)
+		}
+
+		ls.RemoveLocation(lr)
+		ls.AddLocation(newLoc)
+	}
+
+	return toBuilder.String()
+}
+
+func (ls *locSet) RenameLocation(fromLocation, toLocation string) {
```

Comment:
This can call MoveLocation. Or you could remove the function to have a smaller API surface

Reply:
I'm a little confused here. The comment you made makes it sound like it could just call `MoveLocation` but the code shows no changes. I also think it would be advantageous to have one of them call the other (or both call into some more generic code) since they're both doing almost exactly the same thing if you want to keep both functions exported. The only difference in the written code between the two is how `toBuilder` is generated

Reply:
Ah, I meant to say that I don't plan on changing it.  But after revisiting, you're right, and I've updated it.